src/collection/tx_data/collector.py

```diff
@@ -0,0 +1,282 @@
+        self._call_timestamps.append(time.time())
+
+    def run(self):
+        """
+        Main method to process all necessary operations to collect and write data.
+        """
+        raise NotImplementedError("Implement me!")
+
+    async def async_run(self):
+        """
+        Main method to process all necessary operations to collect and write data. Data collection is concurrent.
+        """
+        while not self._collection_completed:
+            self._get_assignment()
+            start_time = time.perf_counter()
+            log_message = "START: function `_get_data`."
+            # Log the start of the function
+            LOG.info(log_message)
+            await self._async_get_data()
+            elapsed_time = time.perf_counter() - start_time
+            LOG.info(f"DONE: function `_get_data` in {elapsed_time:.2f} seconds")
+            self._write_tx_data()
+        LOG.info(f"Collection completed for {self.__class__.__name__}.")
+
+    @property
+    @abstractmethod
+    def COLLECTION_STREAM(self) -> db.CollectionStreamTypes:  # pylint: disable=invalid-name
+        """Implement in subclasses to define the constant value"""
+        raise NotImplementedError("Implement me!")
+
+    def _get_data(self) -> None:
+        """
+        Collect transactions from blocks.
+        Collected transactions are stored in `rel_transactions` attribute.
+        """
+        raise NotImplementedError("Implement me!")
+
+    async def _async_get_data(self):
+        """
+        Concurrently collect transactions from blocks.
+        Collected transactions are stored in `rel_transactions` attribute.
+        """
+        self.rel_transactions.clear()
+        # Use asyncio.gather to fetch blocks concurrently
+        blocks = await asyncio.gather(
+            *(self._async_fetch_block(block_number) for block_number in self.assignment)
+        )
+        for block, block_number in blocks:
+            if not block:
+                continue
+            self._select_relevant_tx_from_block(block, block_number)
+
+    async def _async_fetch_block(self, block_number: int) -> Tuple[UiConfirmedBlock | None, int]:
+        """
+        Use solana client to fetch block with provided number.
+        """
+        await self._async_rate_limit_calls()
+        # Fetch block data.
+        try:
+            block = await self.async_solana_client.get_block(
+                slot=block_number,
+                encoding='jsonParsed',
+                max_supported_transaction_version=0,
+            )
+        except SolanaRpcException as e:
+            LOG.error(f"SolanaRpcException while fetching {block_number}: {e}")
+            await asyncio.sleep(0.5)
+            return await self._async_fetch_block(block_number)
+        except RPCException as e:
+            LOG.error(f"RpcException while fetching {block_number}: {e}")
+            return None, block_number
+        except SerdeJSONError as e:
```

Comment:
What is this kind of an error?

Reply:
This error randomly appears, not were often, when we call some get blocks. If try to get the same block again - it does not persist.